torch/utils/data/dataloader.py

```diff
@@ -598,7 +599,33 @@ def __init__(self, loader: DataLoader) -> None:
         self._timeout = loader.timeout
         self._collate_fn = loader.collate_fn
         self._sampler_iter = iter(self._index_sampler)
-        self._base_seed = torch.empty((), dtype=torch.int64).random_(generator=loader.generator).item()
+
+        # NOTE [RNG re-seeding in Dataloader workers]
+        # When num_workers > 0, we need to ensure that all RNGs are different across workers.
+        # That concerns the builtin random module, numpy RNG, and torch Generators.
+        # We do that by generating a base seed (one per Generator), and we then re-seed the Generators in `_worker_loop()`.
+        # - For the default Generator, we sample its base seed just below (self._base_seed) via the `generator`
+        #   parameter that was passed to `Dataloader(...)`.
+        #   We also use this _base_seed to re-seed the numpy and builtin RNGs.
+        # - For all other Generator objects g, their base seed should only depend on g, not on the `generator` parameter.
+        #
+        # We generate _base_seed for the default Generator here, but for the other Generators, we generate their seed
+        # later in _worker_loop(). Why? Because if we were generating the seeds here, we would have to give _worker_loop()
+        # a Generator -> base_seed mapping, and that is really difficult to do (a simple dict would fail because Generators
+        # can't be pickled, and other solutions get very complex very fast).
```

Comment:
@ezyang thanks a lot for your review so far. Following up on https://github.com/pytorch/pytorch/pull/107131#discussion_r1295854974, do you mind making sure that the above NOTE makes sense to you before I land this?

This is the simplest solution I could come up with.

(In case you're interested, I tried to sample all seeds in the main process instead of in _worker_loop, and my best shot is https://github.com/pytorch/pytorch/commit/e4e3d0a2f24ba4c2f907f430a7dd9058c7fd4ebd - it's reaaaaaally complicated).

Reply:
Comment lgtm